File: queue/number_of_students_unable_to_eat_lunch.cpp
class Solution {
public:
    int countStudents(vector<int>& students, vector<int>& sandwiches) {
        if (students.size() == 1) return students[0] == sandwiches[0] ? 0 : 1;

        queue<int> q;
        stack<int> s;

        for (int i = 0; i < students.size(); ++i) q.push(students[i]);
        for (int i = sandwiches.size() - 1; i >=0; --i) s.push(sandwiches[i]);

        int refused = 0;
        while (!q.empty() && !s.empty()) {
            int student = q.front();
            q.pop();
            if (student == s.top()) {
                s.pop();
                refused = 0;
            } else {
                q.push(student);
                ++refused;
            }

            if (refused > students.size()) return q.size();
        }

        return q.size();
    }
};

File: queue/number_of_students_unable_to_eat_lunch.cs
public class Solution {
    public int CountStudents(int[] students, int[] sandwiches) {
        if (students.Length == 1) return students[0] != sandwiches[0] ? 1 : 0;

        Queue<int> studentsQueue = [];
        Stack<int> sandwichesStack = [];

        for (int i = 0; i < students.Length; ++i) studentsQueue.Enqueue(students[i]);
        for (int i = sandwiches.Length - 1; i >= 0; --i) sandwichesStack.Push(sandwiches[i]);

        int refuse = 0;
        while (studentsQueue.TryDequeue(out int student) && sandwichesStack.TryPeek(out int sandwich))  {
            if (student == sandwich) {
                sandwichesStack.Pop();
                refuse = 0;
            } else {
                studentsQueue.Enqueue(student);
                ++refuse;               
            }

            if (refuse > students.Length) return studentsQueue.Count;
        }

        return studentsQueue.Count;
    }
}
